### PR TITLE
GH-106176, GH-104702: Fix reference leak when importing across multiple threads

### PR DESCRIPTION
**Pull request by brettcannon**
When importing, import deadlock detection tracks which threads are importing which modules in a list. Prior to this change, a list was created per thread but never disposed of. Now, the list is disposed of properly using weakrefs to guarantee GC doesn't interrupt the cleanup.

Co-Authored-By: Kirill Podoprigora <kirill.bast9@mail.ru>

* Issue: gh-106176
* Issue: gh-104702

### CODE REVIEW

**Comment by brettcannon** [comment]
!buildbot .*Windows11.*Refleaks

**Comment by Eclips4** [comment]
This is probably what we should have done.
I'll take a closer look later today (UTC+3).
However, I'd like to hear @exarkun opinion if he has free time.

**Comment by brettcannon** [comment]
FYI the Windows 11 refleak builder has not come back yet, so the green CI is a little misleading until we can get confirmation this PR doesn't leak under Windows.

File: Lib/importlib/_bootstrap.py
Comment (vstinner) [inline Lib/importlib/_bootstrap.py]: I suppose that it cannot be created until _weakref is available. Would you mind to add a comment explaining that? It wasn't obvious to me when I first read the code.

File: Lib/importlib/_bootstrap.py
Comment (vstinner) [inline Lib/importlib/_bootstrap.py]: You may add a reference to the issue, since this issue is complicated.

File: Lib/importlib/_bootstrap.py
Comment (vstinner) [inline Lib/importlib/_bootstrap.py]: You can remove these methods to make the code shorter: ``__getitem__()``, ``__setitem__()``, ``__delitem()``, ``__repr__()``. You may also remove comments to make the code shorter.

**Review by vstinner — COMMENTED** [review]
I'm now curious if it does fix the issue.

**Comment by vstinner** [comment]
I'm still able to reproduce the leak in the main branch:

```
vstinner@WIN C:\victor\python\main>python -m test test_import -m test_concurrency -R 3:3

test_import leaked [38, 36, 40] references, sum=114
test_import leaked [38, 36, 40] memory blocks, sum=114
```

And with this PR: **this is no leak, it works!**

```
vstinner@WIN C:\victor\python\main>python -m test test_import -m test_concurrency -R 3:3
Running Debug|x64 interpreter...
0:00:00 Run tests sequentially
0:00:00 [1/1] test_import
beginning 6 repetitions
123456
......

== Tests result: SUCCESS ==

1 test OK.

Total duration: 7.7 sec
Tests result: SUCCESS
```

Well done @brettcannon and @Eclips4!

**Review by vstinner — APPROVED** [review]
LGTM since it does fix the leak :-) I left a review, feel free to address it, or ignore my remarks ;-) Thanks for fixing Windows Refleaks buildbots!

**Review by Eclips4 — APPROVED** [review]
+1 to all of Victor's words.
LGTM!
(I've check it on my windows setup and this solution seems fine, refleaks are gone)

**Comment by Eclips4** [comment]
I would prefer to backport it to 3.12, because:
1. Original issue was introduced in 3.12
2. This problem looks like one of real situations which can happen.

**Comment by Eclips4** [comment]
This also solves the #104702.

**Comment by brettcannon** [comment]
Thanks for the reviews and verification! I should be able to get this merged at worst by Friday, but hopefully sooner.

**Comment by brettcannon** [comment]
I addressed the review comments from Victor, updated against `main`, flagged this for 3.12, and now I'm waiting for CI to go green again before I merge.

**Review by vstinner — APPROVED** [review]
LGTM. Thanks for the update!

**Comment by vstinner** [comment]
@brettcannon:
> I addressed the review comments from Victor, updated against main, flagged this for 3.12, and now I'm waiting for CI to go green again before I merge.

Thanks. I took the freedom of merging your PR. It became difficult to read buildbot results these days, since there are more and more known failures. This fix should reduce the number of failed buildbots!

**Comment by vstinner** [comment]
Again, thanks for the fix @brettcannon and @Eclips4!

File: Lib/importlib/_bootstrap.py
Comment (Yhg1s) [inline Lib/importlib/_bootstrap.py:56]: FWIW, I think it would be nice to use `__slots__` for this class, to reduce the size of each instance (especially since the setdefault call creates one each time, even when it isn't necessary).

Comment (brettcannon) [reply Lib/importlib/_bootstrap.py:56]: As in you're going to make a PR to make that change, or you're hoping someone else will open an issue on your behalf to track the idea?

Comment (Yhg1s) [reply Lib/importlib/_bootstrap.py:56]: The latter (although not on _my_ behalf) -- I don't know if there's a reason not to do it, and I don't have the spare cycles to dig in and find out.

**Comment by Yhg1s** [comment]
Nevermind, the leak was from a different issue (still not resolved on main, but more cleverly hidden).

**Pull request merged**
